tests/online/cert.c

```diff
@@ -0,0 +1,257 @@
+	g_options.fetch_opts.callbacks.certificate_check = certificate_check_cb;
+
+	g_cert_options.expect_valid = 1;
+	g_cert_options.expect_host = "github.com";
+
+	cl_git_pass(git_clone(&g_repo, "http://github.com/libgit2/TestGitRepository", "./fake", &g_options));
+
+	cl_assert_equal_i(g_cert_options.had_callback, 1);
+}
+
+void test_online_cert__passthrough_https(void)
+{
+	if (!g_has_ssl)
+		cl_skip();
+
+	g_options.fetch_opts.callbacks.certificate_check = certificate_check_cb;
+
+	g_cert_options.expect_valid = 1;
+	g_cert_options.return_code = GIT_PASSTHROUGH;
+
+	cl_git_pass(git_clone(&g_repo, "https://github.com/libgit2/TestGitRepository", "./fake", &g_options));
+}
+
+void test_online_cert__passthrough_https_bad_certificate(void)
+{
+	if (!g_has_ssl)
+		cl_skip();
+
+	g_options.fetch_opts.callbacks.certificate_check = certificate_check_cb;
+
+	g_cert_options.expect_valid = 0;
+	g_cert_options.return_code = GIT_PASSTHROUGH;
+
+	cl_git_fail_with(GIT_ERROR,
+					 git_clone(&g_repo, "https://wrong.host.badssl.com/fake.git", "./fake", &g_options));
+}
+
+void test_online_cert__passthrough_https_other(void)
+{
+	if (!g_has_ssl)
+		cl_skip();
+
+	g_options.fetch_opts.callbacks.certificate_check = certificate_check_cb;
+
+	g_cert_options.expect_valid = 0;
+	g_cert_options.return_code = GIT_EPEEL;
+
+	cl_git_fail_with(GIT_EPEEL,
+					 git_clone(&g_repo, "https://wrong.host.badssl.com/fake.git", "./fake", &g_options));
+}
+
+void test_online_cert__passthrough_ssh(void)
+{
+	if (!g_has_ssh)
+		cl_skip();
+
+	g_options.fetch_opts.callbacks.certificate_check = certificate_check_cb;
+
+	g_cert_options.expect_valid = 0;
+	g_cert_options.return_code = GIT_PASSTHROUGH;
+
+	/* Authentication required, no cb set */
+	cl_git_fail_with(GIT_ERROR,
+					 git_clone(&g_repo, "ssh://github.com/libgit2/TestGitRepository", "./fake", &g_options));
+}
+
+static int cert_check_assert_invalid(git_cert *cert, int valid, const char* host, void *payload)
+{
+	GIT_UNUSED(cert); GIT_UNUSED(host); GIT_UNUSED(payload);
+
+	cl_assert_equal_i(0, valid);
+
+	return GIT_ECERTIFICATE;
+}
+
+void test_online_cert__badssl_expired(void)
+{
+	git_clone_options opts = GIT_CLONE_OPTIONS_INIT;
+	opts.fetch_opts.callbacks.certificate_check = cert_check_assert_invalid;
+
+	if (!g_has_ssl)
+		cl_skip();
+
+	cl_git_fail_with(GIT_ECERTIFICATE,
+			 git_clone(&g_repo, "https://expired.badssl.com/fake.git", "./fake", NULL));
+
+	cl_git_fail_with(GIT_ECERTIFICATE,
+			 git_clone(&g_repo, "https://expired.badssl.com/fake.git", "./fake", &opts));
+}
+
+void test_online_cert__badssl_wrong_host(void)
+{
+	git_clone_options opts = GIT_CLONE_OPTIONS_INIT;
+	opts.fetch_opts.callbacks.certificate_check = cert_check_assert_invalid;
+
+	if (!g_has_ssl)
+		cl_skip();
+
+	cl_git_fail_with(GIT_ECERTIFICATE,
+			 git_clone(&g_repo, "https://wrong.host.badssl.com/fake.git", "./fake", NULL));
+	cl_git_fail_with(GIT_ECERTIFICATE,
+			 git_clone(&g_repo, "https://wrong.host.badssl.com/fake.git", "./fake", &opts));
```

Comment:
It's the same here as above. Adding the callback here doesn't add anything to any of these tests which are about making sure we do reject a bunch of common certificate errors.

Reply:
Adding the callback is there to check we don't have a difference in behavior w.r.t invalid certificates whether there's a callback or not. I can remove if you feel like it's too much though.